File: admin_app/admin_flutter/lib/models/cash_closing.dart
class CashClosing {
  const CashClosing({
    required this.id,
    required this.status,
    required this.closedAt,
    required this.openingAmount,
    required this.expectedCashAmount,
    required this.countedCashAmount,
    required this.cashDifferenceAmount,
    required this.totalSalesAmount,
    required this.totalSalesCount,
    required this.totalWithdrawalAmount,
    required this.totalSupplyAmount,
    this.authorizedByOperatorId,
    this.authorizedByOperatorName,
    required this.payments,
    required this.movements,
    this.number,
    this.cashRegisterNumber,
    this.operatorName,
    this.openedAt,
    this.notes,
    this.treasuryNotes,
    this.treasuryCheckedAt,
    this.businessDate,
    this.crossedBusinessDay = false,
    this.businessDayCutoffMinutes = 180,
  });

  final int id;
  final String? number;
  final String? cashRegisterNumber;
  final String? operatorName;
  final DateTime? openedAt;
  final DateTime closedAt;
  final double openingAmount;
  final double expectedCashAmount;
  final double countedCashAmount;
  final double cashDifferenceAmount;
  final double totalSalesAmount;
  final int totalSalesCount;
  final double totalWithdrawalAmount;
  final double totalSupplyAmount;
  final int? authorizedByOperatorId;
  final String? authorizedByOperatorName;
  final String status;
  final String? notes;
  final String? treasuryNotes;
  final DateTime? treasuryCheckedAt;
  final DateTime? businessDate;
  final bool crossedBusinessDay;
  final int businessDayCutoffMinutes;
  final List<CashClosingPayment> payments;
  final List<CashClosingMovement> movements;

  factory CashClosing.fromJson(Map<String, dynamic> json) {
    return CashClosing(
      id: json['id'] as int,
      number: json['number'] as String?,
      cashRegisterNumber: json['cash_register_number'] as String?,
      operatorName: json['operator_name'] as String?,
      openedAt: json['opened_at'] == null
          ? null
          : DateTime.parse(json['opened_at'] as String).toLocal(),
      closedAt: DateTime.parse(json['closed_at'] as String).toLocal(),
      openingAmount: _toDouble(json['opening_amount']),
      expectedCashAmount: _toDouble(json['expected_cash_amount']),
      countedCashAmount: _toDouble(json['counted_cash_amount']),
      cashDifferenceAmount: _toDouble(json['cash_difference_amount']),
      totalSalesAmount: _toDouble(json['total_sales_amount']),
      totalSalesCount: json['total_sales_count'] as int? ?? 0,
      totalWithdrawalAmount: _toDouble(json['total_withdrawal_amount']),
      totalSupplyAmount: _toDouble(json['total_supply_amount']),
      authorizedByOperatorId: json['authorized_by_operator_id'] as int?,
      authorizedByOperatorName: json['authorized_by_operator_name'] as String?,
      status: json['status'] as String? ?? 'pending_treasury',
      notes: json['notes'] as String?,
      treasuryNotes: json['treasury_notes'] as String?,
      treasuryCheckedAt: json['treasury_checked_at'] == null
          ? null
          : DateTime.parse(json['treasury_checked_at'] as String).toLocal(),
      businessDate: json['business_date'] == null
          ? null
          : DateTime.parse(json['business_date'] as String),
      crossedBusinessDay: json['crossed_business_day'] as bool? ?? false,
      businessDayCutoffMinutes:
          json['business_day_cutoff_minutes'] as int? ?? 180,
      payments: ((json['payments'] as List<dynamic>?) ?? [])
          .map(
            (item) => CashClosingPayment.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
      movements: ((json['movements'] as List<dynamic>?) ?? [])
          .map(
            (item) =>
                CashClosingMovement.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
    );
  }
}

class CashClosingReviewPayload {
  const CashClosingReviewPayload({
    required this.status,
    this.notes,
    this.countedCashAmount,
  });

  final String status;
  final String? notes;
  final double? countedCashAmount;

  Map<String, dynamic> toJson() {
    return {
      'status': status,
      'notes': notes?.trim().isEmpty == true ? null : notes?.trim(),
      if (countedCashAmount != null) 'counted_cash_amount': countedCashAmount,
    };
  }
}

class CashClosingPayload {
  const CashClosingPayload({
    required this.operatorName,
    this.cashRegisterNumber,
    required this.openedAt,
    required this.openingAmount,
    required this.expectedCashAmount,
    required this.countedCashAmount,
    required this.totalSalesAmount,
    required this.totalSalesCount,
    required this.totalWithdrawalAmount,
    required this.totalSupplyAmount,
    required this.payments,
    required this.movements,
    this.authorizedByOperatorId,
    this.authorizedByOperatorName,
    this.notes,
  });

  final String? operatorName;
  final String? cashRegisterNumber;
  final DateTime? openedAt;
  final double openingAmount;
  final double expectedCashAmount;
  final double countedCashAmount;
  final double totalSalesAmount;
  final int totalSalesCount;
  final double totalWithdrawalAmount;
  final double totalSupplyAmount;
  final int? authorizedByOperatorId;
  final String? authorizedByOperatorName;
  final String? notes;
  final List<CashClosingPaymentPayload> payments;
  final List<CashClosingMovementPayload> movements;

  Map<String, dynamic> toJson() {
    return {
      'operator_name': operatorName,
      'cash_register_number': cashRegisterNumber,
      'opened_at': openedAt?.toIso8601String(),
      'opening_amount': openingAmount,
      'expected_cash_amount': expectedCashAmount,
      'counted_cash_amount': countedCashAmount,
      'total_sales_amount': totalSalesAmount,
      'total_sales_count': totalSalesCount,
      'total_withdrawal_amount': totalWithdrawalAmount,
      'total_supply_amount': totalSupplyAmount,
      'authorized_by_operator_id': authorizedByOperatorId,
      'authorized_by_operator_name': authorizedByOperatorName,
      'notes': notes?.trim().isEmpty == true ? null : notes?.trim(),
      'payments': payments.map((payment) => payment.toJson()).toList(),
      'movements': movements.map((movement) => movement.toJson()).toList(),
    };
  }
}

class CashClosingPayment {
  const CashClosingPayment({
    required this.id,
    required this.method,
    required this.amount,
  });

  final int id;
  final String method;
  final double amount;

  factory CashClosingPayment.fromJson(Map<String, dynamic> json) {
    return CashClosingPayment(
      id: json['id'] as int,
      method: json['method'] as String,
      amount: _toDouble(json['amount']),
    );
  }
}

class CashClosingPaymentPayload {
  const CashClosingPaymentPayload({required this.method, required this.amount});

  final String method;
  final double amount;

  Map<String, dynamic> toJson() => {'method': method, 'amount': amount};
}

class CashClosingMovement {
  const CashClosingMovement({
    required this.id,
    required this.movementType,
    required this.amount,
    this.reason,
    this.createdAt,
    this.authorizedByOperatorId,
    this.authorizedByOperatorName,
  });

  final int id;
  final String movementType;
  final double amount;
  final String? reason;
  final DateTime? createdAt;
  final int? authorizedByOperatorId;
  final String? authorizedByOperatorName;

  factory CashClosingMovement.fromJson(Map<String, dynamic> json) {
    return CashClosingMovement(
      id: json['id'] as int,
      movementType: json['movement_type'] as String,
      amount: _toDouble(json['amount']),
      reason: json['reason'] as String?,
      createdAt: json['created_at'] == null
          ? null
          : DateTime.parse(json['created_at'] as String).toLocal(),
      authorizedByOperatorId: json['authorized_by_operator_id'] as int?,
      authorizedByOperatorName: json['authorized_by_operator_name'] as String?,
    );
  }
}

class CashClosingMovementPayload {
  const CashClosingMovementPayload({
    required this.movementType,
    required this.amount,
    this.reason,
    this.createdAt,
    this.authorizedByOperatorId,
    this.authorizedByOperatorName,
  });

  final String movementType;
  final double amount;
  final String? reason;
  final DateTime? createdAt;
  final int? authorizedByOperatorId;
  final String? authorizedByOperatorName;

  Map<String, dynamic> toJson() {
    return {
      'movement_type': movementType,
      'amount': amount,
      'reason': reason,
      'created_at': createdAt?.toIso8601String(),
      'authorized_by_operator_id': authorizedByOperatorId,
      'authorized_by_operator_name': authorizedByOperatorName,
    };
  }
}

double _toDouble(Object? value) {
  if (value == null) return 0;
  if (value is num) return value.toDouble();
  return double.tryParse(value.toString()) ?? 0;
}

File: backend/app/models/cash_closing.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Integer, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class CashClosing(Base):
    __tablename__ = "cash_closings"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    number: Mapped[Optional[str]] = mapped_column(String(40), unique=True, index=True)
    cash_session_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("pdv_cash_sessions.id", ondelete="SET NULL"),
        index=True,
    )
    cash_register_number: Mapped[Optional[str]] = mapped_column(String(10), index=True)
    operator_name: Mapped[Optional[str]] = mapped_column(String(150))
    opened_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    closed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        index=True,
    )
    business_date: Mapped[Optional[date]] = mapped_column(Date)
    crossed_business_day: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
    )
    business_day_cutoff_minutes: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=180,
    )
    opened_by_user_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    closed_by_user_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    opening_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    expected_cash_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    counted_cash_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    cash_difference_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    total_sales_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    total_sales_count: Mapped[int] = mapped_column(nullable=False, default=0)
    total_withdrawal_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    total_supply_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    authorized_by_operator_id: Mapped[Optional[int]] = mapped_column()
    authorized_by_operator_name: Mapped[Optional[str]] = mapped_column(String(150))
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="pending_treasury")
    treasury_checked_by_user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL")
    )
    treasury_checked_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    treasury_notes: Mapped[Optional[str]] = mapped_column(Text)
    notes: Mapped[Optional[str]] = mapped_column(Text)

    payments = relationship(
        "CashClosingPayment",
        back_populates="closing",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    movements = relationship(
        "CashClosingMovement",
        back_populates="closing",
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class CashClosingPayment(Base):
    __tablename__ = "cash_closing_payments"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    closing_id: Mapped[int] = mapped_column(ForeignKey("cash_closings.id", ondelete="CASCADE"))
    method: Mapped[str] = mapped_column(String(30), nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)

    closing = relationship("CashClosing", back_populates="payments")


class CashClosingMovement(Base):
    __tablename__ = "cash_closing_movements"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    closing_id: Mapped[int] = mapped_column(ForeignKey("cash_closings.id", ondelete="CASCADE"))
    movement_type: Mapped[str] = mapped_column(String(30), nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    reason: Mapped[Optional[str]] = mapped_column(String(220))
    created_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    authorized_by_operator_id: Mapped[Optional[int]] = mapped_column()
    authorized_by_operator_name: Mapped[Optional[str]] = mapped_column(String(150))

    closing = relationship("CashClosing", back_populates="movements")
